controllers/etcd_controller_test.go

```diff
@@ -54,85 +90,229 @@ var _ = Describe("Druid", func() {
+			Expect(err).NotTo(HaveOccurred())
 		})
 		AfterEach(func() {
 			c.Delete(context.TODO(), instance)
 		})
 	})
-	Context("when adding etcd resources with statefulset already present", func() {
-		var err error
-		var instance *druidv1alpha1.Etcd
-		var c client.Client
-		var ss *appsv1.StatefulSet
-		BeforeEach(func() {
-			instance = getEtcd("foo3", "default")
 
-			// Setup the Manager and Controller.  Wrap the Controller Reconcile function so it writes each request to a
-			// channel when it is finished.
+	DescribeTable("when adding etcd resources with statefulset already present",
+		func(name string, setupStatefulSet StatefulSetInitializer, setupPod bool) {
+			var ss *appsv1.StatefulSet
+			instance := getEtcd(name, "default", false)
+			c := mgr.GetClient()
+			switch setupStatefulSet {
+			case WithoutOwnerReference:
+				ss = createStatefulset(name, "default", instance.Spec.Labels)
+				Expect(ss.OwnerReferences).Should(BeNil())
+			case WithOwnerReference:
+				ss = createStatefulsetWithOwnerReference(instance)
+				Expect(len(ss.OwnerReferences)).ShouldNot(BeZero())
+			default:
+				Fail("StatefulSetInitializer invalid")
+			}
 
-			Expect(err).NotTo(HaveOccurred())
-			c = mgr.GetClient()
-			ss = createStatefulset("foo3", "default", instance.Spec.Labels)
 			storeSecret := instance.Spec.Backup.Store.SecretRef.Name
-			tlsClientSecret := instance.Spec.Etcd.TLS.ClientTLSSecretRef.Name
-			tlsServerName := instance.Spec.Etcd.TLS.ServerTLSSecretRef.Name
-			errors := createSecrets(c, instance.Namespace, storeSecret, tlsClientSecret, tlsServerName)
+			errors := createSecrets(c, instance.Namespace, storeSecret)
 			Expect(len(errors)).Should(BeZero())
 			c.Create(context.TODO(), ss)
-		})
-		It("should adopt statefulset ", func() {
 			defer WithWd("..")()
-			Expect(ss.OwnerReferences).Should(BeNil())
-			err = c.Create(context.TODO(), instance)
-
+			err := c.Create(context.TODO(), instance)
 			Expect(err).NotTo(HaveOccurred())
-
 			s := &appsv1.StatefulSet{}
-			getStatefulset(c, instance, s)
-			testLog.Info("fetched Statefulset", "statefulset", s.Name)
-			Eventually(len(s.OwnerReferences), timeout).ShouldNot(BeZero())
-		})
-		AfterEach(func() {
-			c.Delete(context.TODO(), instance)
+			Eventually(func() error { return statefulsetIsCorrectlyReconcilced(c, instance, s) }, timeout, pollingInterval).Should(BeNil())
+			setStatefulSetReady(s)
+			err = c.Status().Update(context.TODO(), s)
+			Expect(err).NotTo(HaveOccurred())
+			err = c.Delete(context.TODO(), instance)
+			Expect(err).NotTo(HaveOccurred())
+		},
+		Entry("when statefulset not owned by etcd, druid should adopt the statefulset", "foo20", WithoutOwnerReference, false),
+		Entry("when statefulset owned by etcd with owner reference set, druid should remove ownerref and add annotations", "foo21", WithOwnerReference, false),
+		Entry("when etcd has the spec changed, druid should reconcile statefulset", "foo3", WithoutOwnerReference, false),
```

Comment:
Should we have a case for pre-existing `StatefulSet` with annotations? I know this won't happen when we make the next release. But it would be good to cover the case which would happen for every reconciliation after that. WDYT?

Reply:
Pondering upon how the test would be like, what would be the expected behavior of the test. That the annotations still persist after the reconcile?

Reply:
Yes. Basically, the same `StatefulSet` is adopted and left unmodified (or updated if spec changed). That would be the usual case, right?

Reply:
Yup. Handled that.

Reply:
This change looks good.